### DISCUSSION TITLE
How to properly close CliWrap

### DISCUSSION THREAD

**Discussion by Sadness96** [Tyrrrz/CliWrap#146] [Q&A]
Hi.
I have a problem，Although I have seen many people ask similar questions, I still can't solve them，I want to know how to close a program that runs indefinitely,be similar to ping 0.0.0.0 -t.
`
var cmd = Cli.Wrap("ping").WithArguments(new[] { "baidu.com", "-t" });
`
I tried to follow [Timeout and cancellation](https://github.com/Tyrrrz/CliWrap#timeout-and-cancellation) operation,But it doesn't work.

**Comment by Tyrrrz** [comment]
Passing a cancellation token is the right way. What exactly doesn't work?

**Reply by Sadness96** [reply]
I use WPF to create two buttons,Call CliWrap when clicking start After, clicking stop,Passing a cancellation token, Ping is still running
```
private Command _cmd { get; set; }

public void Start()
{
	_cmd = Cli.Wrap("ping").WithArguments(new[] { "baidu.com", "-t" });
	Task.Run(async () =>
	{
		await foreach (var cmdEvent in _cmd.ListenAsync())
		{
			switch (cmdEvent)
			{
				case StartedCommandEvent started:
					Console.WriteLine($"Process started; ID: {started.ProcessId}");
					break;
				case StandardOutputCommandEvent stdOut:
					Console.WriteLine($"Out> {stdOut.Text}");
					break;
				case StandardErrorCommandEvent stdErr:
					Console.WriteLine($"Err> {stdErr.Text}");
					break;
				case ExitedCommandEvent exited:
					Console.WriteLine($"Process exited; Code: {exited.ExitCode}");
					break;
			}
		}
	});
}

public void Stop()
{
	using var cts = new CancellationTokenSource();
	cts.CancelAfter(TimeSpan.FromSeconds(10));
	var result = _cmd.Observe(cts.Token);
}
```

**Reply by Tyrrrz** [reply]
You run it in

```
		await foreach (var cmdEvent in _cmd.ListenAsync())
```

without cancellation token

**Reply by Sadness96** [reply]
I need to run it to ensure that the console is always output asynchronously
`await foreach (var cmdEvent in _cmd.ListenAsync())`
I can't stop because of it.
Excuse me, how can I ensure that I can correctly click the button to trigger the stop in the case of output?
Is the output changed to WithStandardOutputPipe?

**Reply by Tyrrrz** [reply]
You need to pass cancelation token to that method: `_cmd.ListenAsync(cts.Token)`.
CancellationTokenSource needs to be created before the operation is started. Then if you need to start another operation after that, you should dispose it and create a new source again.

**Reply by Sadness96** [reply]
I see. Thank you very much！
In other words, the end time can only be set during initialization, Cannot end at any time during the run

**Reply by Tyrrrz** [reply]
Well not so much the time, but the token. You can trigger the token by timeout (via `cts.CancelAfter()`) or you can do it manually (via `cts.Cancel()`). In the second approach you have more control.
